{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing data from fasta files\n",
    "\n",
    "for record in SeqIO.parse(\"SARS-CoV-2.fasta\", \"fasta\"):\n",
    "    SARS = record.seq\n",
    "\n",
    "for record in SeqIO.parse(\"Figure1Figure2_032321.fasta\", \"fasta\"):\n",
    "    if 'Moderna' in record.id:\n",
    "        moderna = record.seq\n",
    "    else:\n",
    "        pfizer = record.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from DNA to RNA\n",
    "mrna_SARS = SARS.transcribe()\n",
    "mrna_moderna = moderna.transcribe()\n",
    "mrna_pfizer = pfizer.transcribe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a function that takes mRNA as an input and returns the translated coding region (Hint 1: you need to find the initiation codon and the stop codon. Hint 2: use BioPython!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def coding_region(mrna):\n",
    "    start = 'ATG'\n",
    "    stop = ['TAG','TAA','TGA']\n",
    "\n",
    "    start_index = mrna.find(start)+3\n",
    "    catch = np.arange(start_index, len(mrna), 3)\n",
    "    \n",
    "    for i in catch:\n",
    "        codon = mrna[i : i+3]\n",
    "        if codon in stop:\n",
    "            stop_index = i\n",
    "            break\n",
    "            \n",
    "    return mrna[start_index:stop_index]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. What’s the position of the initiation codon (“AUG”)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The possition of ATG in pfizer is: 54\n",
      "The possition of ATG in moderna is: 57\n",
      "The possition of ATG in SARS is: 106\n"
     ]
    }
   ],
   "source": [
    "pfizer_start = pfizer.find(\"ATG\")\n",
    "moderna_start = moderna.find(\"ATG\")\n",
    "SARS_start = SARS.find(\"ATG\")\n",
    "                             \n",
    "print(f'The possition of ATG in pfizer is: {pfizer_start}')\n",
    "print(f'The possition of ATG in moderna is: {moderna_start}')\n",
    "print(f'The possition of ATG in SARS is: {SARS_start}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. What’s the position of the termination codon?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The possition of STOP in pfizer is: 3873\n",
      "The possition of STOP in moderna is: 3876\n",
      "The possition of STOP in SARS is: 133\n"
     ]
    }
   ],
   "source": [
    "pfizer_end = pfizer_start+len(coding_region(pfizer))+3\n",
    "moderna_end = moderna_start+len(coding_region(moderna))+3\n",
    "SARS_end = SARS_start+len(coding_region(SARS))+3\n",
    "\n",
    "print(f'The possition of STOP in pfizer is: {pfizer_end}')\n",
    "print(f'The possition of STOP in moderna is: {moderna_end}')\n",
    "print(f'The possition of STOP in SARS is: {SARS_end}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. What sequence did each company use for the stop codon?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pfizer used TGATGA as stop codon\n",
      "moderna used TGATAATAG as stop codon\n"
     ]
    }
   ],
   "source": [
    "stop = ['TAG','TAA','TGA']\n",
    "\n",
    "# finding the stop codon length for pfizer\n",
    "pfizer_counter = 0\n",
    "\n",
    "for i in range(pfizer_end, len(pfizer),3):\n",
    "    if pfizer[i : i+3] in stop:\n",
    "        pfizer_counter+=3\n",
    "    else:\n",
    "        break\n",
    "\n",
    "# finding the stop codon length for moderna\n",
    "moderna_counter= 0\n",
    "\n",
    "for i in range(moderna_end, len(moderna),3):\n",
    "    if moderna[i : i+3] in stop:\n",
    "        moderna_counter+=3\n",
    "    else:\n",
    "        break\n",
    "        \n",
    "print(f'pfizer used {pfizer[pfizer_end:pfizer_end+pfizer_counter]} as stop codon')\n",
    "print(f'moderna used {moderna[moderna_end:moderna_end+moderna_counter]} as stop codon')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Please suggest another stop codon sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alternative stop codon : TAATAGTAG\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "\n",
    "alter_stop = ''\n",
    "num_of_end_codons = 3 # to be selected by whom \n",
    "for codon in range (num_of_end_codons):\n",
    "    alter_stop += random.choice(stop)\n",
    "        \n",
    "print(\"Alternative stop codon : \" + str(alter_stop))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. How would you call the area between the first nucleotide and the “AUG” site?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " The area between the first nucleotide and the “AUG” site is called: Putative 5’ UTR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6. How would you call the area between the termination codon and the last nucleotide?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The area between the termination codon and the last nucleotide is divided into 2 parts:\n",
    "1. 3’ UTR\n",
    "2. polyA region"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using your function, translate the coding area into protein. What’s the length of each coding sequence?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "pfizer_protein = coding_region(pfizer).translate()\n",
    "moderna_protein = coding_region(moderna).translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The length of pfizer sequence is 1272\n",
      "The length of moderna sequence is 1272\n"
     ]
    }
   ],
   "source": [
    "print(f'The length of pfizer sequence is {len(pfizer_protein)}')\n",
    "print(f'The length of moderna sequence is {len(moderna_protein)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Are the coding regions of the vaccines identical in the protein level?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "print(pfizer_protein==moderna_protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Are the coding regions of the vaccines identical in the nucleotide level?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    }
   ],
   "source": [
    "print(coding_region(pfizer)==coding_region(moderna))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a histogram of amino-acids usage for the Pfizer vaccine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def protein_frequency(seq):\n",
    "    dicto = {}\n",
    "    for n in seq:\n",
    "        keys = dicto.keys()\n",
    "        if n in keys:\n",
    "            dicto[n] += 1\n",
    "        else:\n",
    "            dicto[n] = 1\n",
    "    return dicto"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAEICAYAAACktLTqAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAZjUlEQVR4nO3de7xcZX3v8c+XBJIgYALZhAAhUYlIgoISi6IckUAFQYmtYGgtAfHA8YACYgXRFvCU1tOXIrTK8URu8VIgRW2iR7kY5BTk1kQiEAOGmxAIuUG4iyb8+sfz7HRl2Ju9ZmbP3smT7/v1mteetdY8a/3WXL7zzLNm1lZEYGZmZdlisAswM7P+53A3MyuQw93MrEAOdzOzAjnczcwK5HA3MyuQw30jImmRpAMHu44qSc9LemMvy46TdMtA1zQQJP2dpFWSnuzQ+g+UtLQT6+5hW5+StDw/lju81mO6qZJ0tqRLBruOjYnDvR9IuknS05KGtbOeiJgcETf1U1n9IiK2iYiHBruOgSRpHHAGMCkiduqndYak3ftjXT2s+zhJ63JoPytpoaQj8rItgQuAP82P5eoSH9OI+PuI+ORg17Excbi3SdIE4AAggA8PajHWX8YDqyNiRbMNJQ3tQD113BYR2wAjgUuB2ZK2B8YAw4FFndrwIO6zvQaHe/uOBW4HrgBmVBdIukLSxZJ+lntVv5S0k6QLc0//Pklvr9z+EUkH5+vnSpot6TuSnstDNlMqt90zf2JYk5e95huLpIskPZZ7dgskHVBZNiR/rH0wb2tB7r1u0OPMH+nn5nXcCbypsg5J+rqkFZKekXS3pL16qWX9flb29Xv5+nBJ35O0Ou/bf0gak5cdL2lxrvEhSSc1rPfzkpZJekLSJxtqHybpq5IezUMU35I0oofaDgZuAHbOj9kVef6H8/28Jt/vezbsz5mS7gZeaAw7Sf+er/46r/NjlWVn5PtsmaTjK/Nr1dsoIl4BLgNGALsD9+dFayTdmNcdknaX1L2P3ZcXJa3/ybqkT+T7+2lJ10kaX1kWkk6WtARY0sP9eK2kUxrm/VrSn+XrrT4fJ0u6QdJT+X45O8+vPocm5Ppm5PtvlaQvVta/haSz8vpXK73Otu/rvt3kRIQvbVyAB4D/CewL/BEYU1l2BbAqLxsO3Ag8THpDGAL8HfCLyu0fAQ7O188Ffg98MN/2H4Db87It83bPBrYCDgKeA/Z4jTo/DuwADCUNOTwJDM/L/hq4B9gDELA3sENeFsDu+fpVwGzgdcBewOPALXnZB4AFpJ6jgD2Bsb3Usn4/K/v6vXz9JODHwNZ5v/cFtsvLDie9oQh4H/Ai8I687NC8T5Nz2+821H4hMBfYHtg2b+MfeqnvQGBpZfrNwAvAIfm+/3y+/7eq7M9CYBwwopd1rq+lso21wJfzOj+Y92dUC/UeV3kchgKn5ufD64EJedtDe6ulMv/7wJX5+rS8j3vmdX4JuLVhHTfk+l61z6Tn+C8r05OANcCwVp+P+X5Ylm8/PE/v18NzqHufv016k9sbeBnYMy8/jdQh2xUYBvzf7v0u6TLoBWzKF+C9pEAfnafvA06vLL8C+HZl+tPA4sr0W4E1lelH2DDcf15ZNgl4KV8/IL8YtqgsvxI4t4nanwb2ztfvB47s5XZB6gEOyfv6lsqyv6+EykHAb4F3VevqZZ3r97Oyr90vzE8AtwJvq7EP/wacmq9fRiX8cs3dtYsUzm+qLH838HAv6z2QDcP9b4DZlektSG9sB1b25xN91NpTuL/EhqG7It9/zdZ7HOmNYg2pM3F75Xk0gRrhDpxJenMekad/BpzQsM8vAuMr6zjoNfZ327wP3bc/H7isnecjcAxwVy/tq8+h7n3etbL8TmB6vr4YmFpZNpb03B7aW32b4sXDMu2ZAVwfEavy9L/QMDQDLK9cf6mH6W1eY/3Vb2q8CAzPH/l3Bh6L9BG82++AXWD9t266P2ofkOedkT9iPyNpDalXNzq3HQc82Me+dpF6WY81bBOAiLgR+AbwTWC5pJmStutjnT35LnAdcFUeXvlHpYOCSDpM0u35I/kaUm+3ex92bqiter2L1JtfkIdV1gDX5vl17MyG+/pKXv8uvWyvrtURsbYy/SLp+dBKvbdHxMiIGB0R74qIn9ctQtJhpN7+tIh4Kc8eD1xU2f5TpDedWvscEc8B/w+YnmdNJ30y6N5mK8/HOs/TqsbXT/drbTzwo8q+LQbWkY5PFMPh3qI8/nk08D5JTyp9Ze50YG9Je3d4808A4yRVH7/dSL1JIn3rZpt8uTkH/Jm53lERMRJ4hvRihfQifROvbSWpdziuYZvrRcQ/RcS+pKGRN5M+XvfkBVJ4dVv/jZSI+GNEnBcRk4D9gSOAY5W+ifQD4Kukoa+RwE8r+7CM9DG7W7XOVaQ30sk5AEdGxOsjHYCs4wlSIADp+EJe/+OV2/Tn6VXbrbc2SXsAs4CjI6LxzfGkyvZHRsSIiLi1cpu+9vlK4BhJ7yYNj/wib7PV52Od52kdjwGHNezb8Ih4vM+WmxCHe+umkd7tJwH75MuewM2k8cZOuoMUkJ+XtKXSd+M/RBoT78m2pGBeCQyV9LdAtVd9CfC/JE1U8jZJO1RXEBHrgB8C50raWtIkKp9SJL1T0n65l/0C6XjBul7qWQhMz7VPAT5aWc/7Jb1V0hDgWdLH5XWkYwvD8j6szb3NP62sczZwvNKB5q2Bv63U/gpp/PXrknbM29lF0gd6qa/RbOBwSVPz/p1BGsO99bWbbWA5UOu75f1Qby35k9Uc4EsR0fh7hW8BX5A0Od/29ZKOanITPyW9KX4ZuLrySbPV5+NPgJ0knaZ0wHlbSfs1WVP3vp2vfIBYUpekI1tYz0bN4d66GcDlEfFoRDzZfSENTfylOvj1sIj4A+lrl4eRenkXA8dGxH29NLmONIb6W9Lwwu/Z8CP1BaQAu54UqJeSelqNTiF9tH2SdDzh8sqy7UiB9HTexmpSL7snf0PqgT0NnEcazuq2E3BNrmMx8P9JY6nPAZ/JdT4N/AXpgCMAEfEz4J9IvcMHgNvyopfz3zPz/NslPQv8nHTArk8RcT/pAOA/k+7vDwEfyo9DXecCs/JQwNE1bt9yvU14R17nBZVhvOcBIuJHwP8mDY89C9xLer7VFhEvkzoEB7PhY9zS8zE/Bw4h3f9Pkr6l8/5masouIj13rpf0HOkYRStvEhs15QMKZkVR+qrivaRvZ6zt6/ZmpXHP3Yoh6SOStpI0itTr/LGD3TZXDncryUmkcdwHSeP0nxrccswGj4dlzMwK5J67mVmBNooT/owePTomTJgw2GWYmW1SFixYsCoievxx20YR7hMmTGD+/PmDXYaZ2SZF0u96W+ZhGTOzAjnczcwK5HA3MyuQw93MrEAOdzOzAjnczcwK5HA3MyuQw93MrEAOdzOzAm0Uv1DdVOk89X2jBnGOT9RmZp3nnruZWYEc7mZmBXK4m5kVyOFuZlYgh7uZWYEc7mZmBXK4m5kVyOFuZlYgh7uZWYEc7mZmBXK4m5kVqM9wl3SZpBWS7q3M217SDZKW5L+jKsu+IOkBSfdL+kCnCjczs97V6blfARzaMO8sYF5ETATm5WkkTQKmA5Nzm4slDem3as3MrJY+wz0i/h14qmH2kcCsfH0WMK0y/6qIeDkiHgYeAP6kn2o1M7OaWh1zHxMRywDy3x3z/F2Axyq3W5rnvYqkEyXNlzR/5cqVLZZhZmY96e8Dqj2d4LzHE5hHxMyImBIRU7q6uvq5DDOzzVur/6xjuaSxEbFM0lhgRZ6/FBhXud2uwBPtFGg9a/YfhfifhJhtXlrtuc8FZuTrM4A5lfnTJQ2T9AZgInBneyWamVmz+uy5S7oSOBAYLWkpcA7wFWC2pBOAR4GjACJikaTZwG+AtcDJEbGuQ7WbmVkv+gz3iDiml0VTe7n9+cD57RRlZmbt8T/Itk2KjzWY1ePTD5iZFWiz77m7J2hmJXLP3cysQA53M7MCOdzNzArkcDczK5DD3cysQA53M7MCOdzNzArkcDczK5DD3cysQA53M7MCOdzNzArkcDczK5DD3cysQA53M7MCbfan/LWB5VMsmw0M99zNzArkcDczK5DD3cysQA53M7MCOdzNzArkcDczK5DD3cysQA53M7MCOdzNzArkcDczK5DD3cysQG2dW0bS6cAngQDuAY4HtgauBiYAjwBHR8TTbVVpthnz+XisFS333CXtAnwGmBIRewFDgOnAWcC8iJgIzMvTZmY2gNodlhkKjJA0lNRjfwI4EpiVl88CprW5DTMza1LL4R4RjwNfBR4FlgHPRMT1wJiIWJZvswzYsaf2kk6UNF/S/JUrV7ZahpmZ9aCdYZlRpF76G4CdgddJ+njd9hExMyKmRMSUrq6uVsswM7MetHNA9WDg4YhYCSDph8D+wHJJYyNimaSxwIp+qNNsUDV7UBN8YNMGVztj7o8C75K0tSQBU4HFwFxgRr7NDGBOeyWamVmzWu65R8Qdkq4BfgWsBe4CZgLbALMlnUB6AziqPwo1M7P62vqee0ScA5zTMPtlUi/ezMwGSRH/INs/8jAz25BPP2BmVqAieu5mZlX+NO+eu5lZkRzuZmYFcribmRXI4W5mViCHu5lZgRzuZmYFcribmRXI4W5mViCHu5lZgRzuZmYFcribmRXI4W5mViCHu5lZgRzuZmYFcribmRXI4W5mViCHu5lZgRzuZmYFcribmRXI4W5mViCHu5lZgRzuZmYFcribmRXI4W5mViCHu5lZgRzuZmYFGjrYBdjA03lquk2cEx2oxDrNj/Xmq62eu6SRkq6RdJ+kxZLeLWl7STdIWpL/juqvYs3MrJ52h2UuAq6NiLcAewOLgbOAeRExEZiXp83MbAC1HO6StgP+G3ApQET8ISLWAEcCs/LNZgHT2i3SzMya086Y+xuBlcDlkvYGFgCnAmMiYhlARCyTtGNPjSWdCJwIsNtuu7VRxqar2fFQj4XaQPNzdNPVzrDMUOAdwP+JiLcDL9DEEExEzIyIKRExpaurq40yzMysUTvhvhRYGhF35OlrSGG/XNJYgPx3RXslmplZs1oO94h4EnhM0h551lTgN8BcYEaeNwOY01aFZmbWtHa/5/5p4PuStgIeAo4nvWHMlnQC8ChwVJvbMDOzJrUV7hGxEJjSw6Kp7azXzMza49MPmJkVyOFuZlYgh7uZWYEc7mZmBXK4m5kVyKf8NRsA/hm/DTT33M3MCuRwNzMrkMPdzKxADnczswI53M3MCuRwNzMrkMPdzKxADnczswI53M3MCuRwNzMrkMPdzKxAPreMmVlFs+cBgo3zXEDuuZuZFcjhbmZWIIe7mVmBHO5mZgVyuJuZFcjhbmZWIIe7mVmBHO5mZgVyuJuZFcjhbmZWIJ9+wDYbzf6sfGP8SblZXW333CUNkXSXpJ/k6e0l3SBpSf47qv0yzcysGf0xLHMqsLgyfRYwLyImAvPytJmZDaC2wl3SrsDhwCWV2UcCs/L1WcC0drZhZmbNa3fM/ULg88C2lXljImIZQEQsk7RjTw0lnQicCLDbbru1WYYNJI9dm238Wu65SzoCWBERC1ppHxEzI2JKREzp6upqtQwzM+tBOz339wAflvRBYDiwnaTvAcsljc299rHAiv4o1MzM6mu55x4RX4iIXSNiAjAduDEiPg7MBWbkm80A5rRdpZmZNaUTP2L6CnCIpCXAIXnazMwGUL/8iCkibgJuytdXA1P7Y71mZtYan37AzKxADnczswI53M3MCuRwNzMrkMPdzKxADnczswI53M3MCuRwNzMrkMPdzKxADnczswI53M3MCuRwNzMrkMPdzKxADnczswI53M3MCuRwNzMrkMPdzKxADnczswI53M3MCuRwNzMrkMPdzKxADnczswI53M3MCuRwNzMrkMPdzKxADnczswI53M3MCuRwNzMrkMPdzKxALYe7pHGSfiFpsaRFkk7N87eXdIOkJfnvqP4r18zM6min574WOCMi9gTeBZwsaRJwFjAvIiYC8/K0mZkNoJbDPSKWRcSv8vXngMXALsCRwKx8s1nAtHaLNDOz5vTLmLukCcDbgTuAMRGxDNIbALBjL21OlDRf0vyVK1f2RxlmZpa1He6StgF+AJwWEc/WbRcRMyNiSkRM6erqarcMMzOraCvcJW1JCvbvR8QP8+zlksbm5WOBFe2VaGZmzWrn2zICLgUWR8QFlUVzgRn5+gxgTuvlmZlZK4a20fY9wF8B90hamOedDXwFmC3pBOBR4Kj2SjQzs2a1HO4RcQugXhZPbXW9ZmbWvnZ67mZmvdJ5vfX9ehbnRIcq2Tz59ANmZgVyuJuZFcjhbmZWIIe7mVmBHO5mZgVyuJuZFcjhbmZWIH/P3cysH20s3+93z93MrEDuuZvZRqfZ3i/4F66N3HM3MyuQw93MrEAOdzOzAjnczcwK5HA3MyuQw93MrEAOdzOzAjnczcwK5HA3MyuQw93MrEAOdzOzAjnczcwK5HA3MyuQw93MrEAOdzOzAjnczcwK5HA3MyuQw93MrEAOdzOzAnUs3CUdKul+SQ9IOqtT2zEzs1frSLhLGgJ8EzgMmAQcI2lSJ7ZlZmav1qme+58AD0TEQxHxB+Aq4MgObcvMzBooIvp/pdJHgUMj4pN5+q+A/SLilMptTgROzJN7APf3eyEwGli1ibXdXLftujefbbvu/jM+Irp6WjC0nzfUTT3M2+BdJCJmAjM7tP1UhDQ/IqZsSm0312277s1n2657YHRqWGYpMK4yvSvwRIe2ZWZmDToV7v8BTJT0BklbAdOBuR3alpmZNejIsExErJV0CnAdMAS4LCIWdWJbfWhn2Gew2m6u23bdm8+2XfcA6MgBVTMzG1z+haqZWYEc7mZmBSoy3CWtk7SwcplQs91Nkj7QMO80SRc3se3nm6t2g7bddd8r6V8lbd1k+y9KWiTp7rye/Wq221XSHElLJD0k6RuShjWx3Z0kXSXpQUm/kfRTSW+u2TYkfa0y/TlJ59Zsu0PlMX5S0uOV6a1qtK/e3z+WNLLOdivtP5Lrf0uT7STpFkmHVeYdLenamu3HSPqX/FgtkHSbpI/UbNu9z4sk/VrSZyXVzoHq81vSB/NzZrdm2+bp4yR9o482X5d0WmX6OkmXVKa/JumzrdTfRJvGPKl1OpX83PhuZXqopJWSftJsDS2JiOIuwPMttjsJuLxh3u3AAZ3edmNb4PvAZ5to+27gNmBYnh4N7FyjnYA7gePz9BDgUuCimttV3u7/qMzbp+59BvweeBgYnac/B5zbwn13LvC5Nu7vWcAXm2w/G7i5xXr3AhYDw4HXAUuAN7V4f48HPt3CPu8I/Bw4r9n7DJgKPFin5p62naePA77RR5ujgNn5+hbAAuC2yvLbSD+QbKmGTrXpbgfcBYzI04cBC4GftLK+Zi9F9tzbcA1wRHevNff4dwZuGYRabgZ2b+L2Y4FVEfEyQESsiog6vy04CPh9RFye260DTgeOlbRNjfbvB/4YEd/qnhERCyPi5pp1ryV9i+D0mrfvlNuAXereON837wFOIH3VtykRcS/wY+BM4BzgOxHxYI2mBwF/aLi/fxcR/9xCDStIvxI/RVJPPzzskaQDgG8Dh9esuR2/BPbP1ycD9wLPSRqVX6d7kgJ0Y/Uz4PB8/RjgyoHacKnhPqLyEepHdRtFxGpSL/bQPGs6cHXkt92BImko6V3+niaaXQ+Mk/RbSRdLel/NdpNJvaH1IuJZ4BHqvbns1di+Bd8E/lLS69tcT0uUTnQ3leZ+izENuDYifgs8JekdLWz6POAvSI/1P9ZsMxn4VQvb6lFEPETKgR1rNhkGzAGmRcR9TW6u+rpcCHy5Rn1PAGvz0M/+pDfhO0ifVKcAd0c6f1UnjWgYlvlYE22vAqZLGg68jVT7gOjU6QcG20sRsU+Lba8khfqc/PcT/VZV30bkJz2knvuldRtGxPOS9gUOIPWmr5Z0VkRc0UdT0XBqiMr8ARERz0r6DvAZ4KWB2i7/dX9PIL1B3dBE22OAC/P1q/J0U6EbES9Iupr0sf/lZtp2k/RN4L2k3vw7W1kHzT3WfwRuJX1iObXJ7WzwupR0HCmg+9Lde98fuID0CWt/4JlcS6e1nCcRcXceATgG+Gl/FtWXUnvu7fg3YGruiY2IiH7rJdXwUkTsky+fbrZHEhHrIuKmiDgHOAX48xrNFtHwApO0HTCGeidzWwTs20ydvbiQFBiv64d11dX9oh0PbAWcXKeRpB1IwyOXSHoE+GvgY80MbVS8ki91LQLWf0qIiJNJnzp6PHlUXyS9EVgHrKjZ5BXgaOCdks5uZZstuJUU5m8lDcvcTuq5708K/o3dXOCrDOCQDDjcXyUingduAi5jgB+MdkjaQ9LEyqx9gN/VaDoP2FrSsXk9Q4CvkQ501elF3wgMk/TfK7W8s4lhIQAi4inSAcoTmmnXHyLiGdKnhs9J2rJGk4+SxsjHR8SEiBhHOij83k7Wmd0IDJf0qcq8pr5V1U1SF/At0mNde+gxIl4EjiANpQ3E4/XLvL2ncgfmKWAk//Ulgo3dZcCXI6KZYda2Odx7diWwN+njdrO2lrS0cqn9Na02bQPMyl9FvJv0T1LO7atRflF/BPiopCXAauCViDi/zkYr7Q9R+irkorzdVk4U9zXSt3wGXETcBfyaegdHjwEaj+X8gDR+3lH5/p4GvE/Sw5LuJH3T58yaq+geP15E+qbM9aSx/2breIp0bOpLkjr9vxruIT0vbm+Y90xE9PcpdHvSOOb+lWYaR8TSiLioU8X1xqcfsA1I2p/05vZnEdHugVIzGyQOdzOzAnlYxsysQA53M7MCOdzNzArkcDczK5DD3cysQA53M7MC/Sfop1AiyXrcQQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "hist_data = protein_frequency(pfizer_protein)\n",
    "\n",
    "plt.bar(list(hist_data.keys()), hist_data.values(), color='g')\n",
    "plt.title('Amino-acids usage for the Pfizer vaccine')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create the same histogram but color each codon within the amino-acid (i.e. create a stacked histogram)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def codon_frequency(seq):\n",
    "    dicto = {}\n",
    "    for i in range(0,len(seq),3):\n",
    "        keys = dicto.keys()\n",
    "        if seq[i : i+3] in keys:\n",
    "            dicto[seq[i : i+3]] += 1\n",
    "        else:\n",
    "            dicto[seq[i : i+3]] = 1\n",
    "    return dicto"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def codon_dict_to_protein_dict (dicto):\n",
    "    table = {\n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                 \n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n",
    "    }\n",
    "    dict_protein = {}\n",
    "    for i in range(len(dicto)):\n",
    "        keys = dict_protein.keys()\n",
    "        temp_protein = table[str(list(dicto.keys())[i])]\n",
    "\n",
    "        if temp_protein in keys:\n",
    "            dict_protein[temp_protein][str(list(dicto.keys())[i])] = list(dicto.values())[i]\n",
    "        else:\n",
    "            dict_protein[temp_protein] = {str(list(dicto.keys())[i]) : list(dicto.values())[i]}\n",
    "    return dict_protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "pfizer_seq = coding_region(pfizer)\n",
    "pfizer_codon_dict = codon_frequency(pfizer_seq)\n",
    "data_to_hist = codon_dict_to_protein_dict (pfizer_codon_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYcAAAEICAYAAAC0+DhzAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAfBUlEQVR4nO3debwcVZn/8c8XAoSwBwKENSqogAhiHBFEIcFhVZhRNkcMAgM4UYGJAwEX4jrRn4Ogg8NEBaKsUVAYVAQDUQQBw6JBArIkBiRACGGRRVme3x/nXKx09Xa7+/bte/N9v173lao6Vaee6j5dT51T1R1FBGZmZkUrDXYAZmbWe5wczMysxMnBzMxKnBzMzKzEycHMzEqcHMzMrKRnk4OkP0jafbDjKJL0F0mvrVF2hKRfdzumbpD0RUmPS3pkgOrfXdJDA1F3lX19VNKj+b1cv957OlRJOlXSd7q0r5C0VZ4+W9Jnmthmi/y6r1yjfJqk81uMp+a2knaTdE8r9Q6WwTyvjOh0hZLmADsAG0fEX1utJyK261hQHRIRaw52DN0maXNgCrBlRDzWoToD2Doi7utEfRV1HwF8F3geeAV4APh0RFwpaRXgdGDniPhd3mTYvacR8eVB2u9xTa63iEF43SPieuANzaybL0zPj4jNBjSoHtbRnoOkccBuQADv62TdNmi2BJa2khgkdfzio0m/yYl8XVKimCVpNLARMBL4w0DteBCP2ayjOj2s9GHgJuA8YFKxQNJ5kr4l6We5S3mDpI0lnSFpmaS7Jb2lsP5CSXvm6WmSZkn6nqRn8pDT+MK620iaI+nJXFY3MUk6U9KDkp6WdKuk3QplK+du+f15X7fmq+fKLvT6kq7IddwCvK5QhyR9XdJjkp6S9HtJb6oRy6vHWTjW8/P0SEnnS1qaj+23kjbKZR+RND/H+ICkYyvqPUnSYkkPSzq6IvbVJH1N0qI8xHK2pNWrxLYncA2wSX7PzsvL35df5yfz675NxfGcLOn3wLOVJ0tJv8qTv8t1HlIom5Jfs8WSPlJY3lS8lSLiFeAcYHVgK6BvSOFJSdfmukPSVpL6jrHv77ncw+mL4cj8ei+T9HNJWxbKQtJkSfcC91Z5Ha+S9LGKZb+T9M95utX2uJ2kayQ9kV+XU/PyYhsal+OblF+/xyV9qlD/SpKm5vqXKn3ORtd6TSX9R6FdHVlRdp6kL+bp+ZL2L5SNyPveqRDTiFz2Gkm/zMd3DbBBRb07S7oxt7ffqTDc3GjbinqWG77MbfWTSp/PpyRdovSZWwP4GX9v93/J7WM1pfPVw/nvDEmr1dnfv+rvn9G7JO2Ul9c8X6nOeSWX76J0Hngq/7tLoWyOpC8onVufkXS1pA1yWc1zSU0R0bE/4D7g34C3Ai8CGxXKzgMez2UjgWuBBaSEsjLwReC6wvoLgT3z9DTgBWDfvO5/AjflslXyfk8FVgUmAM8Ab6gT54eA9UnDalOAR4CRuew/gHmk7qdIQ2Tr57IAtsrTFwOzgDWANwF/Bn6dy/YCbiVduQrYBhhbI5ZXj7NwrOfn6WOB/wNG5eN+K7B2LtuP1HAEvBt4Dtgpl+2dj2m7vO33K2I/A7gCGA2slffxnzXi2x14qDD/euBZ4D35tT8pv/6rFo7nDmBzYPUadb4aS2EfLwGfz3Xum49nvRbiPaLwPowAjs/tYR1gXN73iFqxFJZfAFyUpw/Mx7hNrvPTwI0VdVyT4ysdM6mN31CY3xZ4Elit1faYX4fFef2Ref7tVdpQ3zF/m5QkdwD+CmyTy08gXdBtBqwG/G/fcVc5jr2BR0ntfQ3gQpZvV+cBX8zTnwUuKGy7H3B3RUwj8vxvSMN9qwHvyu9XX/ybAktzm1iJ1O6WAmMabdtEW14I3AJskt+7+cBx1dbNyz6fX6sNgTHAjcAXauzrINI54W35fduK1Auve76i/nllNLAMOJzUVg7L833npznA/aTP6Op5fnqjc0nN82QHE8M7SQlhgzx/N3Biofw84NuF+Y8D8wvz2wNPVjtpkhr7Lyo+XM/n6d1IH6aVCuUXAdP6EfsyYIc8fQ9wQL2TWn5xXwTeWCj7cuFNnAD8Edi5GFeNOl89ziof7CNzA3xzE8fwY+D4PH0OhZNnjrkvdpFO7q8rlL8DWNDkB+ozwKzC/Eq5Ae9eOJ4jG8RaLTk8z/In7cfy69ffeI8gJZonSRcjNxXa0TiaSA7AyaTkvnqe/xlwVMUxP0e6D9NXx4Q6x7tWPoa+9b8EnNNOeySdGG6vsX2xDfUd82aF8luAQ/P0fGBioWwsqW2PqFLvOeSTTZ5/PbWTw1akk96oPH8B8NnK9wHYIr9faxTqvbAQ/8nA9yvi+DlpZKLutk205YXAhwrzXwXOrrZuXnY/sG9hfi9gYY19/Zz8eaxYXvN8RePzyuHALRX1/QY4Ik/PId1f6yv7N+CqPN30uaTvr5PDSpOAqyPi8Tx/IRVDS6Srjj7PV5mvd5Oq+KTMc8DI3C3dBHgw0hBCnz+Rrjj6nnrq6xrulpdNyd29pyQ9Sbqq7OuObk5qBPWMITXsByv2CUBEXAv8N3AW8KikGZLWblBnNd8nNbKLczf2q0o3VZG0j6Sb8pDCk6Qrq75j2KQituL0GNLVw625e/kkcFVe3oxNWP5YX8n1b1pjf81aGhEvFeafI7WHVuK9KSLWjYgNImLniPhFs0FI2ofU2zgwIp7Pi7cEzizs/wlS0mrqmCPiGeAnwKF50aGkk2XfPltpj82006LKz0/fZ21L4EeFY5sPvEy6P1Opsl39qco6AER62GA+8F5Jo0j3IC+sUeeyiHi2Rr1bAgf1xZdjfCcpiTXathm1Xpdqlmv7eXqTGuvWen/qna/qnleq7L+4bZ9ax1PzXFJLR5KD0vjvwcC7JT2i9MjjicAOknboxD7qeBjYXFLxWLYgXc0SEdtFxJr57/qcIE7O8a4XEesCT5E+7JDemOXG+apYQrpi2bxin6+KiG9ExFtJQzuvJw0PVPMs6eTXZ+NCHS9GxOciYltgF2B/4MN5nPNS4Gukobt1gZ8WjmExaZigTzHOx0mJeLt8Al03ItaJ5p/Eepj0gQXS/ZVc/58L60STdTWj3XibJukNwEzg4IioTK7HFva/bkSsHhE3FtZpdMwXAYdJegepy39d3mer7bGZdtqMB4F9Ko5tZET8ucq6i6nT5qu4iNTDOQC4K6o/nbYYWC+P81er90FSz6EY3xoRMb2JbdtR7f1cru3nfT1cY/ta70+981Wj80rl/ovb1lXrXFJvm071HA4kXW1sC+yY/7YBrm8UQAfcTDrBniRplXyz6r2ksbtq1iK9AUuAEZI+CxSv6r8DfEHS1kreLGn9YgUR8TJwGTBN0ihJ21LoJUl6m6S358z8LOl+ycs14rkDODTHPh74QKGePSRtr/Q8+NOkLufLpLHK1fIxvJSvdv+xUOcs4CP5xtco0vhvX+yvkMafvy5pw7yfTSXtVSO+SrOA/SRNzMc3hTSGfWP9zZbzKNDUdws6EG9Tcs/uclK3vPK58rOBUyRtl9ddR9JB/dzFT0kf7M8DlxSuHFttj1cCG0s6QelG6VqS3t7PmPqO7UvKN9gljZF0QI11ZwFHSNo2t6vTGtR9MaldfpTqvQYi4k/AXOBzklaV9E7S57fP+aTex15KN+dHKt1Y3qyJbdvxKLC+pHUKyy4CPp1fow1In6ta38f4DvBJSW/N79tW+TWueb5qdF4htaHXS/qg0g3+Q0jn3CsbHUydc0lNnUoOk4BzI2JRRDzS90caWvkXDeDjfRHxN1KXdR/SVea3gA9HxN01Nvk5aQz5j6Qu2Qss3407nfQhuJr0In6XdKVX6WOkLtsjpLHWcwtla5NOaMvyPpaSrvKr+QzpCmMZ8DmW/xBtDPwwxzEf+CVpPPUZ4BM5zmXAB0k3bAGIiJ8B3yBdnd5HGpeEdBKHdKV6H3CTpKeBX9Dk898RcQ/pBuo3Sa/3e4H35vehWdOAmXmY4OAm1m853n7YKdd5ugpPLQFExI+Ar5C65E8Dd5LaW9MifefnMmBPln+PW2qPuQ28h/T6P0J6SmqP/sSUnUlqO1dLeoZ0j6Zqksnt6gzSwyT35X9riojFpLa3C3BJnVU/mPf5BCnhfK9Qx4OknseppAT6IKkXvlKjbduRzx8XAQ/kdroJ6aGZucDvSQ8J3JaXVdv+B6R7SxeS7r38GBjdxPmq5nklIpaSrvinkM4pJwH7F4by66l6Lqm3gfLNChvGlB41vZP0dMxLjdY3M+vZn8+w9kj6p9zVXo901ft/Tgxm1iwnh+HrWFI3/H7S2OJHBzccMxtKPKxkZmYl7jmYmVlJT/xI2AYbbBDjxo0b7DDMzIaUW2+99fGIaPYLrP3SE8lh3LhxzJ07d7DDMDMbUiT19xvhTfOwkpmZlTg5mJlZiZODmZmVODmYmVmJk4OZmZU4OZiZWYmTg5mZlTg5mJlZiZODmZmV9MQ3pHvVQ1Ovr1u+2fTduhSJmVl3uedgZmYlTg5mZlbi5GBmZiVODmZmVuLkYGZmJX5aqY5LFnylbvkU/LSSmQ1P7jmYmVmJk4OZmZU4OZiZWYmTg5mZlTg5mJlZScPkIOkcSY9JurOwbLSkayTdm/9dr1B2iqT7JN0jaa+BCtzMzAZOMz2H84C9K5ZNBWZHxNbA7DyPpG2BQ4Ht8jbfkrRyx6I1M7OuaJgcIuJXwBMViw8AZubpmcCBheUXR8RfI2IBcB/wDx2K1czMuqTVew4bRcRigPzvhnn5psCDhfUeystKJB0jaa6kuUuWLGkxDDMzGwidviGtKsui2ooRMSMixkfE+DFjxnQ4DDMza0erP5/xqKSxEbFY0ljgsbz8IWDzwnqbAQ+3E+CKbvuZ2zdcZ96keV2IxMxWJK32HK4AJuXpScDlheWHSlpN0muArYFb2gvRzMy6rWHPQdJFwO7ABpIeAk4DpgOzJB0FLAIOAoiIP0iaBdwFvARMjoiXByh2MzMbIA2TQ0QcVqNoYo31vwR8qZ2gzMxscPknu23IGDf1Jw3XWTh9vy5EYjb8+eczzMysZNj2HM467tqG60w+e0IXIjEzG3rcczAzs5Jh23OYMGdyE2vNH/A4zMyGIvcczMysxMnBzMxKnBzMzKzEycHMzEqcHMzMrMTJwczMSpwczMysxMnBzMxKnBzMzKzEycHMzEqcHMzMrMTJwczMSobtD+9Z75l97evqlk+ccH+XIjGzRtxzMDOzkmHbczj4lMaHNq8LcZiZDUXuOZiZWcmw7TkMF/MWLBrsEMxsBeSeg5mZlTg5mJlZiZODmZmVODmYmVmJk4OZmZU4OZiZWYmTg5mZlbT1PQdJJwJHA0H6wvFHgFHAJcA4YCFwcEQsaytKM1vO/Ddu03Cdbe6e34VIbLhquecgaVPgE8D4iHgTsDJwKDAVmB0RWwOz87yZmQ0h7X5DegSwuqQXST2Gh4FTgN1z+UxgDnBym/sxswL/dpgNtJZ7DhHxZ+BrwCJgMfBURFwNbBQRi/M6i4ENOxGomZl1TzvDSusBBwCvATYB1pD0oX5sf4ykuZLmLlmypNUwzMxsALQzrLQnsCAilgBIugzYBXhU0tiIWCxpLPBYtY0jYgYwA2D8+PHRRhxmXTVt2rS2ys2GgnYeZV0E7CxplCQBE4H5wBXApLzOJODy9kI0M7Nua7nnEBE3S/ohcBvwEnA7qSewJjBL0lGkBHJQJwI1M7PuaetppYg4DTitYvFfSb0IMzMbonr3P/uZtk6D8qe6E4eZ2QrIP59hZmYlvdtzMDPrku1nbl+3fN6kFe8rhe45mJlZiXsOZv109At+3sKGP/cczMysxMnBzMxKnBzMzKzEycHMzEqcHMzMrMTJwczMSpwczMysxMnBzMxKnBzMzKzEycHMzEqcHMzMrMS/rWQ2BM1bsGiwQ7Bhzj0HMzMrcXIwM7MSDytZ11z/q8Prlk+c0KVAzKwh9xzMzKzEPQezIWjcCxc2XGfhwIcxbPgGf5l7DmZmVuKeg1k/XbLgK3XLp7BblyIxGzjuOZiZWYmTg5mZlTg5mJlZie852JDxwl6bDnYIZisM9xzMzKykZ3sOjZ7jXtidMIaFs467tuE6k8/215NXNP91yP4N15lyyZVdiMR6UVs9B0nrSvqhpLslzZf0DkmjJV0j6d7873qdCtbMzLqj3WGlM4GrIuKNwA7AfGAqMDsitgZm53kzMxtCWk4OktYG3gV8FyAi/hYRTwIHADPzajOBA9sN0szMuqudew6vBZYA50raAbgVOB7YKCIWA0TEYkkbVttY0jHAMQBbbLFFG2H0to2vu6Nu+SN77NilSMw6b9q0aW2VW+9qZ1hpBLAT8D8R8RbgWfoxhBQRMyJifESMHzNmTBthmJlZp7WTHB4CHoqIm/P8D0nJ4lFJYwHyv4+1F6KZmXVby8khIh4BHpT0hrxoInAXcAUwKS+bBFzeVoRmZtZ17X7P4ePABZJWBR4APkJKOLMkHQUsAg5qcx/WpglzJjex1vwBj8PMho62kkNE3AGMr1I0sZ16zcxscPnnM8zMrMTJwczMSpwczMysxMnBzMxKnBzMzKykZ3+y22w488+qWK9zz8HMzEqcHMzMrMTJwczMSpwczMysxMnBzMxK/LSSmbXs6Bf8M2rDlXsOZmZW4uRgZmYlTg5mZlbiew5mg+CCeH+DNe7vShxmtbjnYGZmJe45mJl1wFnHXVu3fPLZE7oUSWe452BmZiVODmZmVuLkYGZmJU4OZmZW4uRgZmYlTg5mZlbi5GBmZiX+noPZCmrkev8+2CFYD3PPwczMSpwczMysxMNKtkJp9BMHMPR+5sBsILTdc5C0sqTbJV2Z50dLukbSvfnf9doP08zMuqkTPYfjgfnA2nl+KjA7IqZLmprnT+7AfszaNmHO5CbWmj/gcZj1urZ6DpI2A/YDvlNYfAAwM0/PBA5sZx9mZtZ97fYczgBOAtYqLNsoIhYDRMRiSRtW21DSMcAxAFtssUWbYdhAe2jq9XXLN5u+W5ciac/BpzRu8vO6EIdZr2u55yBpf+CxiLi1le0jYkZEjI+I8WPGjGk1DDMzGwDt9Bx2Bd4naV9gJLC2pPOBRyWNzb2GscBjnQjUzMy6p+WeQ0ScEhGbRcQ44FDg2oj4EHAFMCmvNgm4vO0ozcysqwbiS3DTgfdIuhd4T543M7MhpCNfgouIOcCcPL0UmNiJes3MbHD45zPMzKzEycHMzEqcHMzMrMTJwczMSpwczMysxMnBzMxK/P85DLAL4v0N1ri/K3GY2cBq/Iu/Q+vXft1zMDOzEicHMzMrcXIwM7MSJwczMytxcjAzsxInBzMzK/GjrAPs+l8dXrd84oQuBWJm1g/uOZiZWYmTg5mZlTg5mJlZiZODmZmVODmYmVmJk4OZmZU4OZiZWYmTg5mZlTg5mJlZiZODmZmVODmYmVmJk4OZmZU4OZiZWYmTg5mZlTg5mJlZScv/n4OkzYHvARsDrwAzIuJMSaOBS4BxwELg4IhY1n6oK6ZxL1zYcJ2FAx+Gma1g2uk5vARMiYhtgJ2ByZK2BaYCsyNia2B2njczsyGk5Z5DRCwGFufpZyTNBzYFDgB2z6vNBOYAJ7cVpVmHzFuwaLBDMBsSOnLPQdI44C3AzcBGOXH0JZANO7EPMzPrnraTg6Q1gUuBEyLi6X5sd4ykuZLmLlmypN0wzMysg9pKDpJWISWGCyLisrz4UUljc/lY4LFq20bEjIgYHxHjx4wZ004YZmbWYS0nB0kCvgvMj4jTC0VXAJPy9CTg8tbDMzOzwdDyDWlgV+BwYJ6kO/KyU4HpwCxJRwGLgIPaC9HMzLqtnaeVfg2oRvHEVus1M7PB107PwcysbRtfd0fd8kf22LFLkViRfz7DzMxKnBzMzKzEycHMzEp8z8HMBtUF8f4Ga9zflThsee45mJlZiZODmZmVODmYmVmJ7zlYUy5Z8JW65VPYrUuRmA1f/3XI/nXLp1xyZZcicc/BzMyqcM/BzIa0cVN/0nCdhdP360Ikw4t7DmZmVuLkYGZmJU4OZmZW4uRgZmYlTg5mZlbi5GBmZiVODmZmVuLkYGZmJU4OZmZW4uRgZmYlTg5mZlbi5GBmZiVODmZmVuLkYGZmJU4OZmZW4uRgZmYlTg5mZlbi5GBmZiVODmZmVjJgyUHS3pLukXSfpKkDtR8zM+u8AUkOklYGzgL2AbYFDpO07UDsy8zMOm+geg7/ANwXEQ9ExN+Ai4EDBmhfZmbWYYqIzlcqfQDYOyKOzvOHA2+PiI8V1jkGOCbPvgG4p0G1GwCPtxFWu9sPpzp6IYZeqaMXYuiVOnohhl6poxdiaKaOLSNiTJv7qGrEQFQKqMqy5bJQRMwAZjRdoTQ3Isa3HFCb2w+nOnohhl6poxdi6JU6eiGGXqmjF2LoVB2tGqhhpYeAzQvzmwEPD9C+zMyswwYqOfwW2FrSayStChwKXDFA+zIzsw4bkGGliHhJ0seAnwMrA+dExB/arLbpIagB2n441dELMfRKHb0QQ6/U0Qsx9EodvRBDp+poyYDckDYzs6HN35A2M7MSJwcbEiQN1JN1ZlaFk8MKQtIoSW/Of6sNdjwtuGWwA+gEST+VNG6w4zBrpCeTg6Qt2tz+vyXt0ql4qtR/wkDVXbGft0nauDD/YUmXS/qGpNFN1rGKpDNIjxefC8wEHuj7vStJb2mijq0k7Vpl+W6SXtfs8bSp2ndneoKkXSWd1eTq5wFXS/qUpFVa3N8BkiYX5m+W9ED++0AL9Y2R1O8vUknarE7Ze/tbXysknZA/J+5ZdlhP3pCWdFtE7JSnL42I9/dz++NJj8+OBS4BLoqIOzoY36KIaJjAJI0EjgO2AuYB342Il/qxn9uAPSPiCUnvIv0MyceBHYFtIqLhiUDSN4BRwIkR8UxetjbwNeBl0jfZX9OgjiuBUyPi9xXLxwOnRUTDE4Gkz9Ypjoj4QoPtHwJOr1NBzbImYlsZODQiLujHNjsCHwQOBhYAl0XEN5vcdg3gs8DewPeBV/rKmjkOSTfkeB/M83cAE4E1gHMjYmITdQg4DfgYKfGuBLwEfDMiPt/kcdwD7BURCyuWHwl8KiIaXjhI+iYVX5AtiohPNNj+a8AuwBuB3wM3AjcAv4mIJxrtv1MknRQRX83TB0XEDwplX46IU5uoo+7j/hHxvvYjbV6vZtviVeJr+7txRJwJnClpS1KSODefqC8CLo6IP3YwvnpmAi8C1/P3HyE8vh/7WbnQwA8BZkTEpcCl+YTQjH2BraNwFRART0v6KOlr+fs0Uce4ysSQ65nbjyGSZ6ssGwUcDawP1E0OpEei16SNHkROipOBTUnfu7mGdHL8JHAHUDc5SHo9qT0dBiwlXXgoIvboZygvkl6P1YC1KCSHJq3alxiyX0fEUmBpTjzNOAHYFXhbRCwAkPRa4H8knRgRX2+ijhOBayTtGxH35jpOISXNdzcZx9zC9OdICatpEfHJvN9VgfGkRHEk8G1JT0ZEwx/8lPQM1ROU0i5i7SZCORT4ap4+BfhBoWxvoGFyAN4BPEg6T93MYPeWI6Ln/oDbqk23WedbgNuBlztQ16Im15tXmB7R32MB7gRG5Om7gXcVy5qs44+tlFWsd18rZXW2WQv4NOmK+yvAhv1pE228b5eThnWOBWaRksMvgR2b3P6VvP5WhWUP9DOGvYG7gOnAqBaPo977cX+TddwObFBl+Rjg9n7EMhG4D3gTcAbpqn29Fo+r6f1W2Xad/Np+AfgFKemc226baSX2yuNo9rhIF0B7ky4qbwe+CGzXrWOo/OvVnsMOkp4mZc7V8zT0L5OTx3T3JmX1iaQP9uea3Lbe1cTqzdRBukIEXv1iYJObveoi4JeSHgeeJ/VAkLQV8FSTddwl6cMR8b3iQkkfAuY3WcdvJf1rRHy7oo6jgFubrIN8n+TfgX8hfQB2iohlzW7e7H7qeG1EbJ9j+Q6p57RF5OG2Jryf1Jauk3QVaZivv3F9Cjgo2vtS6M013o9jaf7G/SoRUfpBt4hY0p97IRExW9IRwBzSkM7EiHih2e0rq+vvBpJmANsBz5Cutm8ETu9Hu+qUqDFdbb56BREvA1cBV+WHRg4D5kj6fDQ5ZNlJPXnPoV2S3kN6YfcjfVguBn4cEdWGNgYyjpf5+3BKX1J5jn4kOUk7k+6dXN0Xfx7eWDMibmti+02By0jJ5VZSQ31bjuWfIuLPTdSxEfAj4G/8PRmMB1bNdTzSRB3/D/hn0jc+z4qIvzTapmL70dHmGHLxXla1+X7UswZwIKmNTSAluh9FxNXtxNeP/W8I/Bj4K9DXBt5KGqY6MCIebaKOmsfe7OtSuIBS3veLpPtY/bqI6+9+K7a5ivTLpXeSEsNvSL3qrp7YCp/14uecPD8yIppKuDkp7EdqW+NIw5/nNPM57bThmhyuAy4ELm33hDJcSJpAusIS8IeImN1CHXuQhg/IdVzbj21fIZ3MXmL5K6mWTiSt6ESyrlLnaOAg4JCImNCpWJvcd997Cv1/P4qvxXJF9ONk1q6KHvoolj+pNnsBJdLrsEv+exPwBOmmdL/uYQwmSTNJsf+MdG/0zkGNZzgmBzNb8eRHa3clJYj9gfUjYt3Bjap5+QKqL2EPygXUcvE4OZjZUCXpE6RksCtpWOsG0tDSDaQHQvr7JJhlvXpD2sysGeOAH5K+x7N4kGMZVtxzMDOzkp78+QwzMxtcTg5mZlbi5GBmZiVODmZmVvL/AbkGHOk+HhaxAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = pd.DataFrame(data_to_hist)\n",
    "df = df.transpose()\n",
    "df.plot(kind=\"bar\", stacked=True, legend = None)\n",
    "plt.title('Amino-acids usage for the Pfizer vaccine divided into codons')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q8"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What’s the length of the Wuhan strain genome?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total length of Wuhan genome is: 29903\n"
     ]
    }
   ],
   "source": [
    "print(f'The total length of Wuhan genome is: {len(SARS)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q9"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Report the nucleotide sequence of the Spike region of the Wuhan strain from the initiation codon to the termination codon (hint: use the information in the NCBI website about the starting and stopping position of Spike). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "According to NCBI website, the spike of Wuhan strain starts at possition 21563 (counter satrst at 1) and ends at 25384 \n",
      "\n",
      "That is mean that the sequence is:\n",
      "ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTGTTAATCTTACAACCAGAACTCAATTACCCCCTGCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAAGTTTTCAGATCCTCAGTTTTACATTCAACTCAGGACTTGTTCTTACCTTTCTTTTCCAATGTTACTTGGTTCCATGCTATACATGTCTCTGGGACCAATGGTACTAAGAGGTTTGATAACCCTGTCCTACCATTTAATGATGGTGTTTATTTTGCTTCCACTGAGAAGTCTAACATAATAAGAGGCTGGATTTTTGGTACTACTTTAGATTCGAAGACCCAGTCCCTACTTATTGTTAATAACGCTACTAATGTTGTTATTAAAGTCTGTGAATTTCAATTTTGTAATGATCCATTTTTGGGTGTTTATTACCACAAAAACAACAAAAGTTGGATGGAAAGTGAGTTCAGAGTTTATTCTAGTGCGAATAATTGCACTTTTGAATATGTCTCTCAGCCTTTTCTTATGGACCTTGAAGGAAAACAGGGTAATTTCAAAAATCTTAGGGAATTTGTGTTTAAGAATATTGATGGTTATTTTAAAATATATTCTAAGCACACGCCTATTAATTTAGTGCGTGATCTCCCTCAGGGTTTTTCGGCTTTAGAACCATTGGTAGATTTGCCAATAGGTATTAACATCACTAGGTTTCAAACTTTACTTGCTTTACATAGAAGTTATTTGACTCCTGGTGATTCTTCTTCAGGTTGGACAGCTGGTGCTGCAGCTTATTATGTGGGTTATCTTCAACCTAGGACTTTTCTATTAAAATATAATGAAAATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCTCAGAAACAAAGTGTACGTTGAAATCCTTCACTGTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACTTGTGCCCTTTTGGTGAAGTTTTTAACGCCACCAGATTTGCATCTGTTTATGCTTGGAACAGGAAGAGAATCAGCAACTGTGTTGCTGATTATTCTGTCCTATATAATTCCGCATCATTTTCCACTTTTAAGTGTTATGGAGTGTCTCCTACTAAATTAAATGATCTCTGCTTTACTAATGTCTATGCAGATTCATTTGTAATTAGAGGTGATGAAGTCAGACAAATCGCTCCAGGGCAAACTGGAAAGATTGCTGATTATAATTATAAATTACCAGATGATTTTACAGGCTGCGTTATAGCTTGGAATTCTAACAATCTTGATTCTAAGGTTGGTGGTAATTATAATTACCTGTATAGATTGTTTAGGAAGTCTAATCTCAAACCTTTTGAGAGAGATATTTCAACTGAAATCTATCAGGCCGGTAGCACACCTTGTAATGGTGTTGAAGGTTTTAATTGTTACTTTCCTTTACAATCATATGGTTTCCAACCCACTAATGGTGTTGGTTACCAACCATACAGAGTAGTAGTACTTTCTTTTGAACTTCTACATGCACCAGCAACTGTTTGTGGACCTAAAAAGTCTACTAATTTGGTTAAAAACAAATGTGTCAATTTCAACTTCAATGGTTTAACAGGCACAGGTGTTCTTACTGAGTCTAACAAAAAGTTTCTGCCTTTCCAACAATTTGGCAGAGACATTGCTGACACTACTGATGCTGTCCGTGATCCACAGACACTTGAGATTCTTGACATTACACCATGTTCTTTTGGTGGTGTCAGTGTTATAACACCAGGAACAAATACTTCTAACCAGGTTGCTGTTCTTTATCAGGATGTTAACTGCACAGAAGTCCCTGTTGCTATTCATGCAGATCAACTTACTCCTACTTGGCGTGTTTATTCTACAGGTTCTAATGTTTTTCAAACACGTGCAGGCTGTTTAATAGGGGCTGAACATGTCAACAACTCATATGAGTGTGACATACCCATTGGTGCAGGTATATGCGCTAGTTATCAGACTCAGACTAATTCTCCTCGGCGGGCACGTAGTGTAGCTAGTCAATCCATCATTGCCTACACTATGTCACTTGGTGCAGAAAATTCAGTTGCTTACTCTAATAACTCTATTGCCATACCCACAAATTTTACTATTAGTGTTACCACAGAAATTCTACCAGTGTCTATGACCAAGACATCAGTAGATTGTACAATGTACATTTGTGGTGATTCAACTGAATGCAGCAATCTTTTGTTGCAATATGGCAGTTTTTGTACACAATTAAACCGTGCTTTAACTGGAATAGCTGTTGAACAAGACAAAAACACCCAAGAAGTTTTTGCACAAGTCAAACAAATTTACAAAACACCACCAATTAAAGATTTTGGTGGTTTTAATTTTTCACAAATATTACCAGATCCATCAAAACCAAGCAAGAGGTCATTTATTGAAGATCTACTTTTCAACAAAGTGACACTTGCAGATGCTGGCTTCATCAAACAATATGGTGATTGCCTTGGTGATATTGCTGCTAGAGACCTCATTTGTGCACAAAAGTTTAACGGCCTTACTGTTTTGCCACCTTTGCTCACAGATGAAATGATTGCTCAATACACTTCTGCACTGTTAGCGGGTACAATCACTTCTGGTTGGACCTTTGGTGCAGGTGCTGCATTACAAATACCATTTGCTATGCAAATGGCTTATAGGTTTAATGGTATTGGAGTTACACAGAATGTTCTCTATGAGAACCAAAAATTGATTGCCAACCAATTTAATAGTGCTATTGGCAAAATTCAAGACTCACTTTCTTCCACAGCAAGTGCACTTGGAAAACTTCAAGATGTGGTCAACCAAAATGCACAAGCTTTAAACACGCTTGTTAAACAACTTAGCTCCAATTTTGGTGCAATTTCAAGTGTTTTAAATGATATCCTTTCACGTCTTGACAAAGTTGAGGCTGAAGTGCAAATTGATAGGTTGATCACAGGCAGACTTCAAAGTTTGCAGACATATGTGACTCAACAATTAATTAGAGCTGCAGAAATCAGAGCTTCTGCTAATCTTGCTGCTACTAAAATGTCAGAGTGTGTACTTGGACAATCAAAAAGAGTTGATTTTTGTGGAAAGGGCTATCATCTTATGTCCTTCCCTCAGTCAGCACCTCATGGTGTAGTCTTCTTGCATGTGACTTATGTCCCTGCACAAGAAAAGAACTTCACAACTGCTCCTGCCATTTGTCATGATGGAAAAGCACACTTTCCTCGTGAAGGTGTCTTTGTTTCAAATGGCACACACTGGTTTGTAACACAAAGGAATTTTTATGAACCACAAATCATTACTACAGACAACACATTTGTGTCTGGTAACTGTGATGTTGTAATAGGAATTGTCAACAACACAGTTTATGATCCTTTGCAACCTGAATTAGACTCATTCAAGGAGGAGTTAGATAAATATTTTAAGAATCATACATCACCAGATGTTGATTTAGGTGACATCTCTGGCATTAATGCTTCAGTTGTAAACATTCAAAAAGAAATTGACCGCCTCAATGAGGTTGCCAAGAATTTAAATGAATCTCTCATCGATCTCCAAGAACTTGGAAAGTATGAGCAGTATATAAAATGGCCATGGTACATTTGGCTAGGTTTTATAGCTGGCTTGATTGCCATAGTAATGGTGACAATTATGCTTTGCTGTATGACCAGTTGCTGTAGTTGTCTCAAGGGCTGTTGTTCTTGTGGATCCTGCTGCAAATTTGATGAAGACGACTCTGAGCCAGTGCTCAAAGGAGTCAAATTACATTACACATAA\n"
     ]
    }
   ],
   "source": [
    "print(f'According to NCBI website, the spike of Wuhan strain starts at possition 21563 (counter satrst at 1) and ends at 25384 \\n')\n",
    "SARS_coding_region = SARS[21562:25384]\n",
    "print(f'That is mean that the sequence is:\\n{SARS_coding_region}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Report the GC content of the Spike region of the Wuhan strain and compare it to the GC content of the Spike region of the Pfizer vaccine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_GC(seq):\n",
    "    counter = 0\n",
    "    for i in seq:\n",
    "        if i == 'C' or i == 'G':\n",
    "            counter+=1\n",
    "    return counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The GC content of the Spike region of the Wuhan strain is: 1425 (without the start codon)\n",
      "The GC content of the Spike region of pfizer strain is: 2176 (without the start codon)\n"
     ]
    }
   ],
   "source": [
    "print(f'The GC content of the Spike region of the Wuhan strain is: {count_GC(SARS_coding_region[3:])} (without the start codon)')\n",
    "print(f'The GC content of the Spike region of pfizer strain is: {count_GC(pfizer_seq)} (without the start codon)')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q11"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Translate the sequence into amino acids and report the sequence of the protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAIHVSGTNGTKRFDNPVLPFNDGVYFASTEKSNIIRGWIFGTTLDSKTQSLLIVNNATNVVIKVCEFQFCNDPFLGVYYHKNNKSWMESEFRVYSSANNCTFEYVSQPFLMDLEGKQGNFKNLREFVFKNIDGYFKIYSKHTPINLVRDLPQGFSALEPLVDLPIGINITRFQTLLALHRSYLTPGDSSSGWTAGAAAYYVGYLQPRTFLLKYNENGTITDAVDCALDPLSETKCTLKSFTVEKGIYQTSNFRVQPTESIVRFPNITNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCYGVSPTKLNDLCFTNVYADSFVIRGDEVRQIAPGQTGKIADYNYKLPDDFTGCVIAWNSNNLDSKVGGNYNYLYRLFRKSNLKPFERDISTEIYQAGSTPCNGVEGFNCYFPLQSYGFQPTNGVGYQPYRVVVLSFELLHAPATVCGPKKSTNLVKNKCVNFNFNGLTGTGVLTESNKKFLPFQQFGRDIADTTDAVRDPQTLEILDITPCSFGGVSVITPGTNTSNQVAVLYQDVNCTEVPVAIHADQLTPTWRVYSTGSNVFQTRAGCLIGAEHVNNSYECDIPIGAGICASYQTQTNSPRRARSVASQSIIAYTMSLGAENSVAYSNNSIAIPTNFTISVTTEILPVSMTKTSVDCTMYICGDSTECSNLLLQYGSFCTQLNRALTGIAVEQDKNTQEVFAQVKQIYKTPPIKDFGGFNFSQILPDPSKPSKRSFIEDLLFNKVTLADAGFIKQYGDCLGDIAARDLICAQKFNGLTVLPPLLTDEMIAQYTSALLAGTITSGWTFGAGAALQIPFAMQMAYRFNGIGVTQNVLYENQKLIANQFNSAIGKIQDSLSSTASALGKLQDVVNQNAQALNTLVKQLSSNFGAISSVLNDILSRLDKVEAEVQIDRLITGRLQSLQTYVTQQLIRAAEIRASANLAATKMSECVLGQSKRVDFCGKGYHLMSFPQSAPHGVVFLHVTYVPAQEKNFTTAPAICHDGKAHFPREGVFVSNGTHWFVTQRNFYEPQIITTDNTFVSGNCDVVIGIVNNTVYDPLQPELDSFKEELDKYFKNHTSPDVDLGDISGINASVVNIQKEIDRLNEVAKNLNESLIDLQELGKYEQYIKWPWYIWLGFIAGLIAIVMVTIMLCCMTSCCSCLKGCCSCGSCCKFDEDDSEPVLKGVKLHYT\n"
     ]
    }
   ],
   "source": [
    "SARS_protein = SARS_coding_region[3:len(SARS_coding_region)-3].translate()\n",
    "print(SARS_protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q12"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare between the Spike region of the Pfizer vaccine to the Spike of the Wuhan strain in the protein level. Are there any differences and where?\n",
    "Bonus: explain the differences!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1272, 1272)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(SARS_protein), len(pfizer_protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There is difference in index 984: pfizer - P, SARS - K\n",
      "There is difference in index 985: pfizer - P, SARS - V\n"
     ]
    }
   ],
   "source": [
    "# finding the indices of difference\n",
    "index_of_diff = []\n",
    "for i in range(len(SARS_protein)):\n",
    "    if SARS_protein[i] != pfizer_protein[i]:\n",
    "        index_of_diff.append(i)\n",
    "        \n",
    "for index in index_of_diff:\n",
    "    print(f'There is difference in index {index}: pfizer - {pfizer_protein[index]}, SARS - {SARS_protein[index]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Explenation:\n",
    "The K and V amino acids there are both replaced by ‘P’\n",
    "(Proline). The spikes are mounted on the virus body (‘the\n",
    "nucleocapsid protein’). But Pfizer vaccine is only generating the\n",
    "spikes itself, because of that we’re not mounting them on any\n",
    "kind of virus body. Proline substitution in just the right\n",
    "place would make proteins take up their ‘pre-fusion’\n",
    "configuration, even without being part of the whole virus - this\n",
    "works because Proline is a very rigid amino acid. It acts as a\n",
    "kind of splint, stabilizing to show to the immune system"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q13"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the amino acids that appear in the vaccine and the Wuhan strain, create a look-up table that maps between the Wuhan strain codon to the Pfizer codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ['TTC', 'TTT'],\n",
       " 'GTT': ['GTG', 'GTC', 'CCT'],\n",
       " 'CTT': ['CTG'],\n",
       " 'TTA': ['CTG', 'CTC'],\n",
       " 'TTG': ['CTG', 'CTC'],\n",
       " 'CCA': ['CCT', 'CCC'],\n",
       " 'CTA': ['CTG'],\n",
       " 'GTC': ['GTG', 'GTT', 'GTC'],\n",
       " 'TCT': ['TCC', 'AGC', 'TCT'],\n",
       " 'AGT': ['AGC', 'TCT'],\n",
       " 'CAG': ['CAG'],\n",
       " 'TGT': ['TGT', 'TGC'],\n",
       " 'AAT': ['AAC', 'AAT'],\n",
       " 'ACA': ['ACC', 'ACA'],\n",
       " 'ACC': ['ACC', 'ACA'],\n",
       " 'AGA': ['AGA', 'CGG', 'AGG'],\n",
       " 'ACT': ['ACA', 'ACC'],\n",
       " 'CAA': ['CAG', 'CAA'],\n",
       " 'CCC': ['CCT', 'CCC'],\n",
       " 'CCT': ['CCA', 'CCC', 'CCT'],\n",
       " 'GCA': ['GCC', 'GCT', 'GCA'],\n",
       " 'TAC': ['TAC', 'TAT'],\n",
       " 'TTC': ['TTT', 'TTC'],\n",
       " 'CGT': ['AGA', 'CGG'],\n",
       " 'GGT': ['GGC', 'GGG', 'GGA', 'GGT'],\n",
       " 'TAT': ['TAC', 'TAT'],\n",
       " 'GAC': ['GAC', 'GAT'],\n",
       " 'AAA': ['AAG', 'AAA', 'CCT'],\n",
       " 'TCC': ['TCC', 'AGC'],\n",
       " 'TCA': ['AGC', 'TCT', 'TCC'],\n",
       " 'CAT': ['CAC', 'CAT'],\n",
       " 'TGG': ['TGG'],\n",
       " 'GCT': ['GCC', 'GCT', 'GCA'],\n",
       " 'ATA': ['ATC', 'ATT'],\n",
       " 'GGG': ['GGC', 'GGA'],\n",
       " 'AAG': ['AAG', 'AAA'],\n",
       " 'AGG': ['AGA', 'CGC', 'CGG'],\n",
       " 'GAT': ['GAC', 'GAT'],\n",
       " 'AAC': ['AAC', 'AAT'],\n",
       " 'GAG': ['GAG', 'GAA'],\n",
       " 'GGC': ['GGC', 'GGA'],\n",
       " 'ATT': ['ATC', 'ATT'],\n",
       " 'TCG': ['AGC', 'TCT'],\n",
       " 'GAA': ['GAG', 'GAA'],\n",
       " 'CAC': ['CAC', 'CAT'],\n",
       " 'ATG': ['ATG'],\n",
       " 'GCG': ['GCC'],\n",
       " 'TGC': ['TGC', 'TGT'],\n",
       " 'GGA': ['GGC', 'GGG', 'GGA'],\n",
       " 'GTG': ['GTG', 'GTC'],\n",
       " 'ACG': ['ACC'],\n",
       " 'CTC': ['CTG'],\n",
       " 'GTA': ['GTG'],\n",
       " 'ATC': ['ATC', 'ATT'],\n",
       " 'GCC': ['GCC', 'GCT'],\n",
       " 'AGC': ['AGC', 'TCC'],\n",
       " 'CTG': ['CTG'],\n",
       " 'CGG': ['CGG', 'AGA'],\n",
       " 'CGC': ['CGG']}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mapping from SARS to pfizer\n",
    "\n",
    "SARS_seq = SARS_coding_region[3:len(SARS_coding_region)-3]\n",
    "pfizer_seq\n",
    "look_up_SARS_to_pfizer = {}\n",
    "for i in range(0,len(pfizer_seq),3):\n",
    "    if SARS_seq[i:i+3] in look_up_SARS_to_pfizer.keys():\n",
    "        if str(pfizer_seq[i:i+3]) not in look_up_SARS_to_pfizer[str(SARS_seq[i:i+3])]:\n",
    "            look_up_SARS_to_pfizer[str(SARS_seq[i:i+3])].append(str(pfizer_seq[i:i+3]))\n",
    "    else:\n",
    "        look_up_SARS_to_pfizer[str(SARS_seq[i:i+3])] = [str(pfizer_seq[i:i+3])]\n",
    "\n",
    "#show the first 5 items in the look-up map\n",
    "look_up_SARS_to_pfizer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q14"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Can you deduce any rule about the look-up table that Pfizer uses internally?\n",
    "Bonus: explain the differences!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dict_values_lst = list(look_up_SARS_to_pfizer.values())\n",
    "# codon_diff = []\n",
    "\n",
    "# for i in range(len(dict_values_lst)):\n",
    "#     if dict_values_lst[i][0] != dict_values_lst[i][1]:\n",
    "#         codon_diff.append(dict_values_lst[i][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are in total 351 similar codons \n",
      "which 307 of them are one-to-many \n",
      "and therefore they could choose another codon but did not\n",
      "\n",
      "[39, 54, 57, 78, 108, 114, 123, 126, 129, 153, 159, 165, 168, 189, 213, 216, 225, 237, 270, 282, 285, 291, 300, 303, 333, 336, 360, 381, 429, 432, 438, 441, 456, 462, 465, 492, 516, 528, 534, 552, 573, 576, 579, 612, 615, 621, 633, 639, 645, 660, 666, 678, 696, 699, 708, 732, 747, 753, 771, 774, 777, 786, 792, 795, 810, 846, 867, 879, 891, 894, 909, 912, 921, 930, 945, 966, 987, 1002, 1026, 1029, 1032, 1041, 1056, 1062, 1068, 1071, 1107, 1128, 1140, 1146, 1167, 1209, 1212, 1245, 1287, 1311, 1344, 1347, 1368, 1374, 1389, 1410, 1413, 1419, 1425, 1431, 1434, 1446, 1461, 1491, 1497, 1518, 1521, 1542, 1551, 1575, 1590, 1602, 1605, 1617, 1620, 1623, 1638, 1656, 1662, 1668, 1680, 1689, 1692, 1728, 1737, 1752, 1842, 1851, 1866, 1875, 1887, 1917, 1923, 1938, 1941, 1965, 1977, 1983, 1989, 2007, 2037, 2040, 2070, 2073, 2076, 2079, 2115, 2124, 2127, 2139, 2163, 2166, 2181, 2190, 2193, 2217, 2229, 2241, 2265, 2313, 2319, 2325, 2328, 2331, 2346, 2361, 2397, 2418, 2433, 2436, 2463, 2466, 2469, 2472, 2475, 2490, 2493, 2496, 2505, 2511, 2526, 2544, 2556, 2559, 2562, 2583, 2595, 2613, 2619, 2637, 2640, 2658, 2664, 2688, 2691, 2724, 2736, 2748, 2751, 2766, 2769, 2790, 2802, 2817, 2820, 2832, 2847, 2850, 2853, 2859, 2874, 2898, 2931, 2934, 2949, 2958, 2967, 2985, 2988, 2991, 2994, 3012, 3018, 3036, 3051, 3057, 3063, 3087, 3090, 3111, 3120, 3129, 3132, 3150, 3153, 3165, 3189, 3195, 3207, 3213, 3219, 3228, 3234, 3252, 3258, 3261, 3264, 3270, 3291, 3309, 3336, 3348, 3351, 3360, 3363, 3369, 3390, 3396, 3414, 3432, 3438, 3444, 3462, 3465, 3474, 3501, 3513, 3528, 3537, 3546, 3558, 3564, 3567, 3570, 3588, 3597, 3600, 3609, 3615, 3639, 3654, 3669, 3672, 3687, 3708, 3714, 3717, 3729, 3732, 3735, 3738, 3744, 3753, 3756, 3771, 3777, 3780, 3786, 3801, 3810, 3813]\n"
     ]
    }
   ],
   "source": [
    "# positions of same codons that has one to many in look up map  \n",
    "indicies = []\n",
    "total_codon_similarity = 0\n",
    "for i in range(0,len(pfizer_seq),3):\n",
    "    if SARS_seq[i:i+3] == pfizer_seq[i:i+3]:\n",
    "        total_codon_similarity +=1\n",
    "        if len(look_up_SARS_to_pfizer[str(SARS_seq[i:i+3])]) == 1:\n",
    "            continue\n",
    "        else:\n",
    "            indicies.append(i)\n",
    "print(f'There are in total {total_codon_similarity} similar codons',\n",
    "      f'\\nwhich {len(indicies)} of them are one-to-many',\n",
    "      f'\\nand therefore they could choose another codon but did not\\n')\n",
    "print(indicies)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is noticable that pfizer's rule for codon selection is to use a codon that contains the maximum C or G nocleotid based, and however correspond to the same protein as in the SARS protein\n",
    "\n",
    "##### Explanation:\n",
    "These codons are stronger and it is harder to separate them, thus it has much more resistance. It is important due to the reason that the vaccine need to be transported and stored for long times, and if it won't be resistant, it can be damaged and will not function."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q15"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Without aligning, develop an algorithm to quickly sift through the 75000 genomes and find all entries that can match the South Afircan variant (hint: mutations occur very rarely. You can assume that positions adjacent to the mutated sites are identical to the Wuhan strain)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['G', 'A', 'A']"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[SARS[23011], SARS[23062], SARS[21800]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sa_var(path_to_file, d):\n",
    "    # d denote for the number of nuc from each side of the \"bad\" nuc\n",
    "    \n",
    "    first = SARS[23011-d:23011] + 'A' + SARS[23012:23012+d]\n",
    "    second = SARS[23062-d:23062] + 'T' + SARS[23063:23063+d]\n",
    "    third = SARS[21800-d:21800] + 'C' + SARS[21801:21801+d]\n",
    "\n",
    "    lst_of_id = []\n",
    "    lst_of_countries = []\n",
    "\n",
    "    with open(path_to_file, mode='r') as handle:\n",
    "        for record in SeqIO.parse(handle, 'fasta'):\n",
    "            identifier = record.id\n",
    "            description = record.description\n",
    "            sequence = record.seq\n",
    "\n",
    "            index = (description.find('human'))\n",
    "            index2 = (description.find('/', index+1))\n",
    "            index3 = (description.find('/', index2+1))\n",
    "\n",
    "            if (sequence.find(first) != -1) and (sequence.find(second) != -1) and (sequence.find(third) != -1):\n",
    "                lst_of_id.append(identifier)\n",
    "                lst_of_countries.append(description[index2+1:index3])\n",
    "    \n",
    "    return lst_of_id, lst_of_countries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_file = 'sequences.fasta'\n",
    "id_lst, countries = sa_var(path_to_file, 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q16"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Report all entries in your collection that you think match the South Afican variant. Where were they collected? Does it make sense?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_lst,countries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "45"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(id_lst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['DEU', 'FRA', 'GHA',\n",
       "       'HG999935.1 |Severe acute respiratory syndrome coronavirus 2 isolate a1244 genome assembly, complete genome: monopartit',\n",
       "       'HG999939.1 |Severe acute respiratory syndrome coronavirus 2 isolate a1245 genome assembly, complete genome: monopartit',\n",
       "       'ITA', 'USA'], dtype='<U118')"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(countries)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The SA variant appeard in Deutschland, France, Ghana, Italy and the USA.\n",
    "Although it was expected that SA would appear as well, it does make sense becuse these countries are highly visited and therefore it is possible.\n",
    "(In the original DB there is only one sequence from South Africa, however it does not appear in our list)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q17"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Develop a method to extract the Spike sequence of one of the entries, translate, and report the protein sequence (hint: the South African spike is not very different than the Wuhan strain and must start with AUG)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MW898265.1'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_lst[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('sequences.fasta', mode='r') as handle:\n",
    "    for record in SeqIO.parse(handle, 'fasta'):\n",
    "        identifier = record.id\n",
    "        if identifier == 'MW898265.1' :\n",
    "            sequence = record.seq\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_spike_sa_var(sequence, d):\n",
    "    look_up_start = SARS_coding_region[:d]\n",
    "    start_index = sequence.find(look_up_start)\n",
    "    stop = ['TAG','TAA','TGA']\n",
    "    catch = np.arange(start_index, len(sequence), 3)\n",
    "    \n",
    "    for i in catch:\n",
    "        codon = sequence[i : i+3]\n",
    "        if codon in stop:\n",
    "            stop_index = i\n",
    "            break\n",
    "            \n",
    "    return sequence[start_index+3:stop_index]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "some_sa_var = find_spike_sa_var(sequence, 15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Q18"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use everything you learned so far, including the look up table of Pfizer and the addition of special nucleotides to design a South African mRNA vaccine. Don’t forget the add the 5’UTR and 3’UTR. Report the vaccine sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### step-by-step solution:\n",
    "\n",
    "1. translate the spike SA variant to protein\n",
    "2. Find if there are any gaps in the legnth between the spike SARS and the spike varient. If exists - fix the gap temporarly\n",
    "3. according to Q12, make sure protein in position 984,985 are the same as SARS\n",
    "4. While having same length - find all diferences between SARS and variant\n",
    "5. while there is no difference in the protein level, take the codon sequence  \n",
    "6. If there is change in the protein level, translate the protein randomly according to the table, and change it, if necessary, according to the look-up map that was created. If there are several options for codon, choose the one with the highest GC content\n",
    "7. add:\n",
    "    * start codons to the spike\n",
    "    * stop codons to the spike\n",
    "    * UTR 5' at the beginning\n",
    "    * UTR 3' and polyA region at the end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There is a gap of 3 between the SARS and the variant\n",
      "The gap possition starts as:240\n",
      "LAL\n",
      "The gaps between SARS and fixed variant are in possitions [78, 213, 240, 241, 258, 259, 260, 267, 300, 301, 302, 415, 482, 499, 612, 699]\n",
      "Is 984 & 985 are the same? True\n"
     ]
    }
   ],
   "source": [
    "codon_to_protein_table = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                 \n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n",
    "}\n",
    "\n",
    "# translate the variant to protein\n",
    "var_translated = some_sa_var.translate()\n",
    "\n",
    "print(f'There is a gap of {len(SARS_protein)-len(var_translated)} between the SARS and the variant')\n",
    "\n",
    "# find the gap possition between SARS to variant in protein level\n",
    "for i in range(len(var_translated)):\n",
    "    if var_translated[i] == SARS_protein[i]:\n",
    "        continue\n",
    "    else:\n",
    "        if var_translated[i+1] == SARS_protein[i+1]:\n",
    "            continue\n",
    "        else:\n",
    "            ap_possition = i\n",
    "            break\n",
    "print(f'The gap possition starts as:{i}')\n",
    "print(SARS_protein[i:i+3])\n",
    "\n",
    "# temorarly fixing to the legnth of SARS with adding the \"missing\" part\n",
    "## fixing the protein seq\n",
    "temp_var = var_translated[:239]+'LAL'+var_translated[239:]\n",
    "## fixing the nuc seq\n",
    "adjusted_var_codon = some_sa_var[:239*3] + 'LAL'*3 + some_sa_var[239*3:]\n",
    "\n",
    "# check for difference between SARS and fixed variant\n",
    "errors = []\n",
    "for j in range(len(temp_var)):\n",
    "    if temp_var[j] == SARS_protein[j]:\n",
    "        continue\n",
    "    else:\n",
    "        errors.append(j)\n",
    "\n",
    "print(f'The gaps between SARS and fixed variant are in possitions {errors}')\n",
    "\n",
    "# check if 984,985 are the same\n",
    "print(f'Is 984 & 985 are the same? {temp_var[984:986]==SARS_protein[984:986]}')\n",
    "\n",
    "# taking pfizer vaccine, however where the variant different from SARS, taking the nuc according to look-up map\n",
    "nuc_errors = [item * 3 for item in errors]\n",
    "nuc_errors.append(len(temp_var)*3)\n",
    "nuc_errors.insert(0,-3)\n",
    "new_vacc = ''\n",
    "for p in range(len(nuc_errors)-1):\n",
    "    new_vacc += pfizer[nuc_errors[p]+3:nuc_errors[p+1]]\n",
    "\n",
    "    if adjusted_var_codon[nuc_errors[p+1]:nuc_errors[p+1]+3] in look_up_SARS_to_pfizer.keys():\n",
    "        map_this_nuc = adjusted_var_codon[nuc_errors[p+1]:nuc_errors[p+1]+3]\n",
    "    else:\n",
    "        map_this_nuc = list(codon_to_protein_table.keys())[list(codon_to_protein_table.values()).index(temp_var[(p+1)//3])]\n",
    "    \n",
    "    temp_lst_map = look_up_SARS_to_pfizer[map_this_nuc]\n",
    "    best_codon_seq = ''\n",
    "    best_GC = 0\n",
    "    for codon in temp_lst_map:\n",
    "        temp_count = count_GC(codon)\n",
    "        if temp_count > best_GC:\n",
    "            best_codon_seq = codon\n",
    "\n",
    "    new_vacc += best_codon_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initiate the start and end as in pfizer vaccine\n",
    "sa_vac_start = pfizer[:pfizer_start+3]\n",
    "sa_vac_end = pfizer[pfizer_end:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GAGAATAAACTAGTATTCTTCTGGTCCCCACAGACTCAGAGAGAACCCGCCACC...GCA')"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#combine Start+coding region+end\n",
    "full_new_vacc = sa_vac_start + new_vacc + sa_vac_end\n",
    "full_new_vacc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
